explorer/interpreter/type_checker.cpp

```diff
@@ -188,6 +189,7 @@ static auto IsType(Nonnull<const Value*> value) -> bool {
     case Value::Kind::TypeOfParameterizedEntityName:
     case Value::Kind::TypeOfMemberName:
     case Value::Kind::TypeOfMixinPseudoType:
+    case Value::Kind::TypeOfNamespaceName:
       // These aren't first-class types, but they are still types.
```

Comment:
Is a namespace name really a type? This seems odd to me, I would've thought it a little more abstract and maybe not reference-able. This is what's implied to me by the fail_use-as_expression.carbon, `  // CHECK:STDERR: COMPILATION ERROR: {{.*}}/explorer/testdata/namespace/fail_use_as_expression.carbon:[[@LINE+1]]: expected `.member_name` after name of namespace N`

But I would expect here instead that `N` _should_ be valid as an expression if `N` is a type, because it's necessary that work for reflection purposes.

Could I, for example, instantiate a template function on a namespace? That'd require `N` as an expression without a member access.

If not, I would've thought namespaces should _not_ be types, period.

Reply:
We're modeling them here as placeholders, just like the name of a generic class that hasn't been given parameters, or the name of a field that hasn't been looked up as a member. Like those other cases, a namespace name is an expression (for example, it can appear on the right-hand side of an `alias`), so it needs a type, even though it's not a first-class expression that can be used in arbitrary contexts, and we use placeholder types for those kinds of "expression that isn't first-class" situations.

Reply:
My leaning would be that the goal should be to either not treat this as a type and make it work in some other way, or treat it as a type more fully and allow it in expressions. Trying to use the existing type support feels like a kludge to me; I don't think it accurately represents what's going on (unlike a placeholder type, I don't think this is a placeholder) and I think that could yield problems in the future as behavior may be expected to diverge more fully. Treating it as a type now obfuscates what the issues are.